src/realtime.c

```diff
@@ -0,0 +1,273 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Patrick Griffis <pgriffis@igalia.com>
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <gio/gio.h>
+
+#include "realtime.h"
+#include "request.h"
+#include "xdp-dbus.h"
+#include "xdp-utils.h"
+
+typedef struct _Realtime Realtime;
+typedef struct _RealtimeClass RealtimeClass;
+
+struct _Realtime
+{
+  XdpRealtimeSkeleton parent_instance;
+  GDBusProxy *rtkit_proxy;
+};
+
+struct _RealtimeClass
+{
+  XdpRealtimeSkeletonClass parent_class;
+};
+
+static Realtime *realtime;
+
+GType realtime_get_type (void) G_GNUC_CONST;
+static void realtime_iface_init (XdpRealtimeIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (Realtime, realtime, XDP_TYPE_REALTIME_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_TYPE_REALTIME, realtime_iface_init));
+
+static gboolean
+map_pid_if_needed (XdpAppInfo *app_info, pid_t *pid, GError **error)
+{
+  if (!xdp_app_info_is_host (app_info))
+  {
+    if (!xdg_app_info_map_pids (app_info, pid, 1, error))
+    {
+      g_prefix_error (error, "Could not map pid: ");
+      g_warning ("Realtime error: %s", (*error)->message);
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+static void
+on_call_ready (GObject      *source_object,
+               GAsyncResult *result,
+               gpointer      user_data)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GDBusMethodInvocation) invocation = g_steal_pointer (&user_data);
+  g_autoptr(GVariant) response = NULL;
+
+  response = g_dbus_proxy_call_finish (G_DBUS_PROXY (source_object),
+                                       result,
+                                       &error);
+
+  if (error)
+    g_dbus_method_invocation_return_gerror (invocation, error);
+  else
+    g_dbus_method_invocation_return_value (invocation, g_variant_new ("()"));
+}
+
+static gboolean
+handle_make_thread_realtime_with_pid (XdpRealtime           *object,
+                                      GDBusMethodInvocation *invocation,
+                                      guint64                process,
+                                      guint64                thread,
+                                      guint32                priority)
+{
+  g_autoptr (GError) error = NULL;
+  Request *request = request_from_invocation (invocation);
+  pid_t pids[1] = { process };
+  
+  if (!map_pid_if_needed (request->app_info, pids, &error))
+  {
+    g_dbus_method_invocation_return_gerror (invocation, error);
+    return TRUE;
+  }
+
+  g_dbus_proxy_call (G_DBUS_PROXY (realtime->rtkit_proxy),
+                     "MakeThreadRealtimeWithPID",
+                     g_variant_new ("(ttu)", pids[0], thread, priority),
+                     G_DBUS_CALL_FLAGS_NONE,
+                     -1,
+                     NULL,
+                     on_call_ready,
+                     g_object_ref (invocation));
+
```

Comment:
I guess we do not expect to ever make this interactive?

I still think it should propably still have a permission to let users deny this request.

Reply:
Its just such a low level detail i didnt think anything would want to expose a permission for it.